### ISSUE TITLE
Twig tags in my MJML file

### ISSUE THREAD

**Issue by gregorydavidjenkins** [mjmlio/mjml#457]
I'm hoping to use MJML to create some email templates for my website powered by Craft. Craft uses Twig as its templating language. Instead of coding static content into my MJML templates I'm hoping to leave in my Twig tags.

My issue is this. When I use twig's output tags like `{{ siteurl }}` it works as aspected. But when I use logic tags like `{% set posts = craft.entries.section('blog') %}` it seems to strip those out completely. I need to escape these tags so that they stay in.

Is there anything I can configure to make these tags not get removed?

**Comment by iRyusa** [comment]
Hi @gregorydavidjenkins ,

As mentioned in #459, Twig should be applied prior to MJML, this is the best way to do it. I know that you can change Twig delimiters, it could be a workaround too : http://stackoverflow.com/questions/21794497/use-custom-delimiters-in-the-current-twig-template

Like you mention, some non html-friendly chars are behaving weirdly with our parser. We're working on a new parser for MJML4 but it takes some time to implement it (#394).

**Comment by gregorydavidjenkins** [comment]
Twig before MJML isn't an option because I'm using grunt-mjml to create email templates that are dynamically sent out at an automatic interval. Twig has to persist because Twig injects the correct content into the template at send time.

I did find that wrapping my twig tags in <mj-raw> worked well.

`
<mj-raw>
    {% set time = now|date_modify("-14 days") %}
    {% set posts = craft.entries.section('blog').after(time).limit(3) %}
</mj-raw>`

**Comment by iRyusa** [comment]
Well if you compile your Twig into MJML markup and then make php to compile MJML into HTML it will be the same and you don't have to compile your template through grunt-mjml so it's an easier process, and MJML template are always up-to-date. 

I know this is doable in Rails env, dunno how to do it in PHP ( cf https://github.com/kolybasov/mjml-ruby gem )

**Comment by AlexandreGagner** [comment]
Hi,
Compiling Twig to MJML and compile the MJML to HTML is ok for email such static email or newsletter, but for transactional email or email with dynamic content for each adresse, the twig render is ok, but the MJML render take few seconds to compile and slow down the email sending process 🐌

**Comment by iRyusa** [comment]
Hi,

@AlexandreGagner spawning the node process is really slow. We're looking to speed this process up, but I think the best way is to spawn a cluster of node that communicate through http or socket with your application. We’re also working on a MJML API which would enable you to hand out the MJML compilation, would you be interested in joining the beta?

**Comment by rogierslag** [comment]
You might be interested in https://github.com/inventid/maily which offers a react based mjml http service

**Comment by AlexandreGagner** [comment]
@rogierslag Thanks
@iRyusa Why not, but I'm not sure I will make a lot of feedback (for now, but in a future side project why not)

**Comment by gregorydavidjenkins** [comment]
@iRyusa I'm interested in the beta for sure.

**Comment by ngarnier** [comment]
@gregorydavidjenkins @AlexandreGagner please shoot me an email at ngarnier@mailjet.com, I'll send you invites for the beta when it's ready :)

**Comment by iRyusa** [comment]
Hi there !

Issue looks fixed for now from @AlexandreGagner edit ! So I'm closing it 👍

**Comment by ninze** [comment]
@iRyusa  Are there any side effects of processing mjml before twig? Take `for` loops for example:

we can wrap `{% for %}` tags in `<mj-raw>` (even automatically) which will actually get mjml to output a valid twig template, however what we will be looping inside `for` are HTML fragments, not MJML fragments. In theory mjml will not be aware of the  loop context and also will not be aware of the siblings that those generated elements may have. I am not sure if it it makes any difference for mjml compiler or not. For example, the compiler might yield a different table structure depending on whether there is only 1 row or 10 rows. Just a guess.

Processing mjml before twig makes a lot of sense for performance: mjml only has to be compiled once instead of (possibly) hundreds of thousands of times a day for transactional email.

**Comment by iRyusa** [comment]
Well MJML has a lot of things to do before output the HTML ( add head style for the MJML, compute some size for mj-image, optimize outlook conditionnals, ... ), so loops can produce a invalid html output.
Let's say you loop over columns,  MJML won't be able to know that you do so, and will think that's a 100% column, after the loop you'll have 5 columns of 100% each, browser will behave "fine" but desktop Outlook will display all the column in a single row.

**Comment by ninze** [comment]
@iRyusa Are you saying using mjml as a template preprocessor is generally a bad idea and should be avoided, or is it still OK as long as the end user knows about the side effects?

I found a few packages online that (I think) follow the path of `mjml` => `twig` => `html`:

 - https://github.com/tdeNL/tde-webpack-mjml-plugin (outputs twig templates in the example)
 - https://github.com/assoconnect/mjml-bundle (appears to process mjml at twig compile time and cache result)
 - ... probably a zillion others in other languages / frameworks

In order not to spawn a node instance, there must be a process/server which is running all the time and listening to requests. Is such a server a part of mjml or do I need to search for a 3rd party package?

thanks for your time

**Comment by ninze** [comment]
Having thought about it more, I came to a conclusion that it isn't a great idea to process mjml before template rendering :) A lot of twig features can make things unpredictable, such as `if...else`, `include`, `embed`, etc. For example:

	{% if condition %}
		render some 2 column content
	{% elseif another_condition %}
		render some 3 column content
	{% else %}
		{% include ('another_file.html.twig') %}
	{% endif %}

Mjml would see all of that mess as one content block (plus `mj-raw` tags) and I am pretty sure it wasn't designed for that. In the end I would not be able to use most of the best twig features and need to stick with simple variables, which is not the goal.

I tried installing [mjml-server](https://www.npmjs.com/package/mjml-server), and indeed for an average-sized email, mjml renders in 15ms or so, which is quite OK. Not sure if that minimal http server code is good enough to be used in production though. I would be much more confident to use such a feature if it was part of the main mjml package.
